Error handling in shell scripts

Unrud opened the pull request:
Most errors in the shell scripts for building, packaging and launching are ignored. (e.g. `make install` succeeds even when compiling fails).

This PR tries to mitigate the problem by always enabling `errexit`.

In bash scripts `pipefail` is enabled. In plain sh scripts I removed pipes or checked the output.

I removed some of the subshells used in program arguments, as errors are ignored even with `errexit`.

Besides the main focus of this PR, I added some missing quotes around path variables and replaced an instance of`*` with `./*` for globbing (preventing paths from being interpreted as options).

Unrud committed:
Scripts: Quote path variables

Unrud committed:
Scripts: Prevent paths interpreted as args

Unrud committed:
Scripts: Set errexit to catch errors

Unrud committed:
Sciprts: Set pipefail or remove pipes

Errors in pipes are ignored otherwise

Unrud committed:
Scripts: Remove some subshells that ignore errors

Unrud committed:
Scripts: Remove exit stmts covered by errexit

anvilvapre commented on launch-game.sh:
```diff
@@ -1,15 +1,17 @@
 
 if command -v python3 >/dev/null 2>&1; then
-	 LAUNCHPATH=$(python3 -c "import os; print(os.path.realpath('$0'))")
+	 LAUNCHPATH=$(python3 -c 'import os,sys; print(os.path.realpath(sys.argv[1]))' "$0")
```
is/was LAUNCHPATH ever correct as it seems to include the basename `launch-game.sh`? or was it meant to be a file and not a dir.

alternative: `LAUNCHPATH=$(dirname $(reallink -f "$0"))` or `LAUNCHPATH=$(reallink -f "$0")`

with no python dependency.

anvilvapre replied:
bash -x on old script indeed shows `LAUNCHPATH=/home/../OpenRA/launch-game.sh`

pchote replied:
`LAUNCHPATH` is supposed to contain the `launch-game.sh`. Also, `readlink -f` is a non-portable GNU-ism!

anvilvapre commented on launch-game.sh:
```diff
@@ -1,15 +1,17 @@
 else
-	 LAUNCHPATH=$(python -c "import os; print(os.path.realpath('$0'))")
+	 LAUNCHPATH=$(python -c 'import os,sys; print(os.path.realpath(sys.argv[1]))' "$0")
 fi
```
Similar.

anvilvapre commented on launch-game.sh:
```diff
@@ -32,10 +34,10 @@ then
 
 # Show a crash dialog if something went wrong
-if [ $? != 0 ] && [ $? != 1 ]; then
+if [ $rc != 0 ] && [ $rc != 1 ]; then
```
trivial: "${rc}" like evereywhere

Unrud replied:
Fixed

anvilvapre commented on packaging/functions.sh:
```diff
@@ -140,8 +144,9 @@ install_data() {
+install_windows_launcher() (
+	set -o errexit || exit $?
+
 	SRC_PATH="${1}"
```
optional: sanity check to see if called from correct dir.

anvilvapre commented on packaging/functions.sh:
```diff
@@ -150,14 +155,14 @@ install_windows_launcher()
 	# NET 6 is unable to customize the application host for windows when compiling from Linux,
 	# so we must patch the properties we need in the PE header.
 	# Setting the application icon requires an external tool, so is left to the calling code
 	python3 "${SRC_PATH}/packaging/windows/fixlauncher.py" "${DEST_PATH}/${LAUNCHER_NAME}.exe"
```
trivial: check if dotnet and python3 are present

Unrud replied:
Why? It just fails if they are not available.

anvilvapre commented on packaging/functions.sh:
```diff
@@ -263,14 +274,14 @@ install_linux_shortcuts() {
 		shift
 	done
-}
+)
```
how could this have been missing. or did you add the start as well?

Unrud replied:
I modified the start as well.

anvilvapre commented on packaging/linux/AppRun.in:
```diff
@@ -1,6 +1,8 @@
+HERE=$(dirname "$LAUNCHER")
 
 # Run the game or server
 if [ -n "$1" ] && [ "$1" = "--server" ]; then
```
trivial: redundant check to see if not empty.

anvilvapre commented on packaging/linux/buildpackage.sh:
```diff
@@ -44,9 +45,9 @@ fi
 echo "Downloading appimagetool"
 if command -v curl >/dev/null 2>&1; then
-	curl -s -L -O https://github.com/AppImage/AppImageKit/releases/download/continuous/appimagetool-x86_64.AppImage || exit 3
+	curl -s -L -O https://github.com/AppImage/AppImageKit/releases/download/continuous/appimagetool-x86_64.AppImage
```
quoted "http:// ...."

anvilvapre commented on packaging/linux/buildpackage.sh:
```diff
@@ -44,9 +45,9 @@ fi
+	curl -s -L -O https://github.com/AppImage/AppImageKit/releases/download/continuous/appimagetool-x86_64.AppImage
 else
-	wget -cq https://github.com/AppImage/AppImageKit/releases/download/continuous/appimagetool-x86_64.AppImage || exit 3
+	wget -cq https://github.com/AppImage/AppImageKit/releases/download/continuous/appimagetool-x86_64.AppImage
```
trivial: set url once above

anvilvapre commented on packaging/linux/openra.appimage.in:
```diff
@@ -1,14 +1,17 @@
 	LAUNCHER=${APPIMAGE}
 
 	APPIMAGEID=$(printf "file://%s" "${APPIMAGE}" | md5sum | cut -d' ' -f1)
+	test -n "${APPIMAGEID}"
```
trivial: move to line 12

anvilvapre commented on packaging/linux/openra.appimage.in:
```diff
@@ -43,21 +46,23 @@ fi
 # Show a crash dialog if something went wrong
-if [ $? != 0 ] && [ $? != 1 ]; then
+if [ $rc != 0 ] && [ $rc != 1 ]; then
 	LOGS="${XDG_CONFIG_HOME:-${HOME}/.config}/openra/Logs"
```
side note: storing logs and maps in `.config` better would be to use a user data dir,

abcdefg30 replied:
See #10374. We should not change that in this PR.

anvilvapre commented on packaging/linux/openra.appimage.in:
```diff
@@ -43,21 +46,23 @@ fi
 	fi
 
-	test -d Support/Logs && LOGS="${PWD}/Support/Logs"
+	if [ -d Support/Logs ]; then
```
trivial: executable .. `-x` over `-d`

also: `"./Support/Logs"`.

is this a way to override the log location? for developers, or on macos?

Unrud replied:
> executable .. `-x` over `-d`

But it has to be a directory, right? What would be the point of this?

anvilvapre replied:
> > executable .. `-x` over `-d`
> 
> But it has to be a directory, right? What would be the point of this?

trivial. even it its a directory it won't do you much good if you do not have executable rights on it.

Unrud replied:
> even it its a directory it won't do you much good if you do not have executable rights on it.

It doesn't seem meaningful. Using `-d` is good enough. Checking with `-x` still doesn't mean that you can actually create (or write existing) log files and it has the additional caveat that it succeeds for executable files.

anvilvapre commented on packaging/linux/openra.in:
```diff
@@ -9,21 +11,23 @@ if [ "${1#${PROTOCOL_PREFIX}}" != "${1}" ]; then
 	fi
 
-	test -d Support/Logs && LOGS="${PWD}/Support/Logs"
+	if [ -d Support/Logs ]; then
```
` "./Support/Logs" `

anvilvapre commented on packaging/macos/buildpackage.sh:
```diff
@@ -28,7 +29,7 @@ if [[ "${OSTYPE}" != "darwin"* ]]; then
 
 # Set the working dir to the location of this script
-cd "$(dirname "${0}")" || exit 1
+cd "$(dirname "${0}")"
```
trivial: `pushd`

Unrud replied:
shellcheck says that pushd is undefined in POSIX sh

anvilvapre commented on packaging/package-all.sh:
```diff
@@ -10,7 +12,7 @@ export GIT_TAG="$1"
 export BUILD_OUTPUT_DIR="$2"
 
```
check if dir exists.

anvilvapre commented on packaging/windows/buildpackage.sh:
```diff
@@ -34,9 +34,9 @@ elif [[ ${TAG} == playtest* ]]; then
 
 if command -v curl >/dev/null 2>&1; then
-	curl -s -L -O https://github.com/electron/rcedit/releases/download/v1.1.1/rcedit-x64.exe || exit 3
+	curl -s -L -O https://github.com/electron/rcedit/releases/download/v1.1.1/rcedit-x64.exe
```
`"http:..."`

anvilvapre commented on packaging/windows/buildpackage.sh:
```diff
@@ -34,9 +34,9 @@ elif [[ ${TAG} == playtest* ]]; then
+	curl -s -L -O https://github.com/electron/rcedit/releases/download/v1.1.1/rcedit-x64.exe
 else
-	wget -cq https://github.com/electron/rcedit/releases/download/v1.1.1/rcedit-x64.exe || exit 3
+	wget -cq https://github.com/electron/rcedit/releases/download/v1.1.1/rcedit-x64.exe
```
`"https:.."`

anvilvapre reviewed (dismissed):
please consider proposed changes.

Unrud commented:
> please consider proposed changes.

I've changed variables from `$abc` to `${abc}` and changed the remaining occurrences of `cd "$(dirname "$0")"` to 
```
HERE=$(dirname "$0")
cd "${HERE}"
```

I agree with the rest of the suggestions, which I have not commented on, but I think they are beyond the scope of this PR. I would rather create separate PRs as I have found other problems with the scripts that I would like to fix.

anvilvapre commented:
> > please consider proposed changes.
> 
> I've changed variables from `$abc` to `${abc}` and changed the remaining occurrences of `cd "$(dirname "$0")"` to
> 
> ```
> HERE=$(dirname "$0")
> cd "${HERE}"
> ```
> 
> I agree with the rest of the suggestions, which I have not commented on, but I think they are beyond the scope of this PR. I would rather create separate PRs as I have found other problems with the scripts that I would like to fix.

sure.

pchote commented on launch-dedicated.sh:
```diff
@@ -7,7 +7,7 @@
 
 ENGINEDIR=$(dirname "$0")
-if command -v mono >/dev/null 2>&1 && [ "$(grep -c .NETCoreApp,Version= ${ENGINEDIR}/bin/OpenRA.Server.dll)" = "0" ]; then
+if command -v mono >/dev/null 2>&1 && [ "$(grep -c .NETCoreApp,Version= "${ENGINEDIR}/bin/OpenRA.Server.dll")" = "0" ]; then
```
The outer grep command is already double-quoted. Will this work as expected?

Unrud replied:
Yes, nested quotes are already used in parameter expansions:

https://github.com/OpenRA/OpenRA/blob/c15af9f68abe070cf825fa1dcbf19921b8e8ce1d/launch-dedicated.sh#L16-L16

anvilvapre replied:
Nesting does not work I suspect. You concat 3 strings here. `${Name:-`, `Dedicated Server` and `}`. I.e. if you add `\? ` in there bash will expand it to a questionmark.

pchote commented on packaging/package-all.sh:
```diff
@@ -19,14 +21,12 @@ function build_package() (
 	./buildpackage.sh "${GIT_TAG}" "${BUILD_OUTPUT_DIR}"
 )
 
 #exit on any non-zero exited (failed) command
```
Should this comment be removed too?

Unrud replied:
Removed it.

Unrud committed:
Scripts: Remove obsolete comment

Unrud committed:
Scripts: Check exit status of background process

anvilvapre commented:
b.t.w. on my platform gtk-dialog does not work with `python` (2.7) but does with `python3`. likely because one of the imports fails (Gtk?).

not sure if it's worth checking to see if `python3` exists.

Unrud commented:
> not sure if it's worth checking to see if `python3` exists.

Right. I think it should just default to **python3**. **python2** is end-of-life.

Unrud commented:
I've added a commit to change it to **python3** in https://github.com/OpenRA/OpenRA/pull/20072.

Unrud commented:
Rebased after merge of  #20072

anvilvapre reviewed (approved):
Reviewed in the past.

Mailaender commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/f063f326cf3c464bfe2fcff9c1ac8e683a73a0f5...935000e0f0c7d70aad4859f4d67fec90dab32651)
